src/chapter3/01-a-http-primer.md

```diff
@@ -0,0 +1,226 @@
+Among other things it supports HTTP and is available on both Windows and most Linux distributions.
+Open a command line and run:
+
+```shell
+curl http://localhost:3000/
+```
+
+The will output the following:
+
+```
+Hello, world!
+```
+
+### HTTP URLs
+
+The string `http://localhost:3000/` is a so called URL.
+You have probably seen URLs before.
+Navigate to Google and search for "MERN stack" - you will notice that the URL in your browser looks like this:
+
+```
+https://www.google.com/search?q=mern+stack
+```
+
+For didactic reasons we will include the port:
+
+```
+https://www.google.com:443/search?q=mern+stack
+```
+
+Generally speaking, a typical HTTP URL has the following form:
+
+```
+scheme://host:port/path?key1=value1&key2=value2#test
+```
+
+> Note the word _typical_ here.
+> HTTP URLs (and especially more general URLs) can become _much more_ complicated, but we will not cover all the little details in this book.
+> In fact the complexity of URLs is often a source of subtle bugs and browser crashes.
+> For example - at some point Android Chrome would [crash](https://news.ycombinator.com/item?id=28639708) when trying to open `http://../foo`.
+
+The **scheme** specifies the scheme used and usually indicates the protocol.
+We will almost exclusively use HTTP or HTTPS (which is just secure HTTP) in this book.
+Therefore the scheme will almost always either be http or https.
+In the above Google URL the scheme is clearly https.
+
+You already learned about the **host** and the **port** - the host identifies the device you wish to connect to and the port is the communication endpoint on that device.
+Note that the host could be a domain name (like www.google.com) or an IP address (like 142.251.36.238).
+Usually we will work with domain names since they are stable (unlike a lot of IP addresses).
+Together the host and the port make up the **authority**.
+
+> At the time of this writing 142.251.36.238 is one of Google's IP addresses.
+> This may of course change by the time you are reading this book.
+
+The next part is the **path**.
+Assuming it is not empty, the path begins with a forward slash ("/") and uniquely identifies the resource we want to query.
+In the Google URL the path is "/search".
+Often paths will be hierarchical.
+In this case the different components of the hierarchy are generally separated by slashes - for example "/path/to/resource".
+
+The path can be followed by a **query**.
+The query begins with a question mark and is followed by key-value pairs.
+In the Google URL this is "?q=mern+stack".
+Here the query provides information about your search.
+If there are multiple key-value pairs, they are separated by ampersands ("&").
+For example the query could be "?key1=value1&key2=value2".
+
+The query can be followed by a **fragment**.
+This is used for navigation by the client and is not sent to server.
+We will return to fragment a bit later in the book.
```

Comment:
will we? unless we build a blog or sth I don't really se a use for them

Reply:
we probably will, but this sentence is a bit useless anyway, so should strike it